static/code/common.js

```diff
@@ -0,0 +1,362 @@
+            type: '[',
+            valid: true
+          };
+        } else if (raw.trim() === ']') {
+          split[j] = {
+            type: ']',
+            valid: true
+          };
+        } else {
+          split[j] = {
+            type: 'unparsed'
+          };
+        }
+        split[j].raw = raw;
+        split[j].index = index;
+      }
+      // Replace token with split array.
+      split.unshift(i, 1);
+      Array.prototype.splice.apply(tokens, split);
+    }
+  }
+
+  // Third step is to parse each 'unparsed' token as a number, if it was
+  // preceded by a '[' token.
+  for (var i = 1; i < tokens.length; i++) {
+    var token = tokens[i];
+    if (tokens[i - 1].type === '[' && token.type === 'unparsed') {
+      token.type = 'num';
+      token.value = NaN;
+      token.valid = false;
+      // Does not support E-notation or NaN.
+      if (/^\s*[-+]?(\d*\.?\d*|Infinity)\s*$/.test(token.raw)) {
+        token.value = Number(token.raw);
+        token.valid = !isNaN(token.value);
+      }
+    }
+  }
+
+  // Fourth step is to split remaining 'unparsed' tokens into 'id' and '.'
+  // tokens.
+  var unicodeRegex = /\\u([0-9A-F]{4})/ig;
+  function decodeUnicode(m, p1) {
+    return String.fromCodePoint(parseInt(p1, 16));
+  }
+  for (var i = tokens.length - 1; i >= 0; i--) {
+    var token = tokens[i];
+    if (token.type === 'unparsed') {
+      var index = token.index + token.raw.length;
+      // Split string on periods.
+      var split = token.raw.split(/(\s*\.\s*)/);
+      for (var j = split.length - 1; j >= 0; j--) {
+        var raw = split[j];
+        index -= raw.length;
+        if (raw === '') {
+          split.splice(j, 1);  // Delete the empty string.
+          continue;
+        } else if (raw.trim() === '.') {
+          split[j] = {
+            type: '.',
+            valid: true
+          };
+        } else {
+          // Parse Unicode escapes in identifiers.
+          var valid = true;
+          var value = split[j];
+          while (true) {
+            var test = value.replace(unicodeRegex, '');
+            if (test.indexOf('\\') === -1) {
+              break;
+            }
+            // Invalid escape found.  Trim off last char and try again.
+            value = value.substring(0, value.length - 1);
+            valid = false;
+          }
+          // Decode Unicode.
+          value = value.replace(unicodeRegex, decodeUnicode);
+          split[j] = {
+            type: 'id',
+            value: value,
+            valid: valid
+          };
+        }
+        split[j].raw = raw;
+        split[j].index = index;
+      }
+      // Replace token with split array.
+      split.unshift(i, 1);
+      Array.prototype.splice.apply(tokens, split);
+    }
+  }
+
+  // Finally, validate order of tokens.  Only check for permanent errors.
+  // E.g. '$..foo' can never be legal.
+  // E.g. '$["foo' isn't legal now, but could become legal after more typing.
+  var state = 0;
```

Comment:
Again, use named states: `need id`, `got id`, `got [`, `need ]`,  or similar.

Reply:
It's more complicated than that, and will get much more complicated.